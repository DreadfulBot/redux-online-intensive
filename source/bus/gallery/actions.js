import { createAction } from "redux-actions";
export const showNextPhoto = createAction("SHOW_NEXT_PHOTO");
export const showPrevPhoto = createAction("SHOW_PREV_PHOTO");
export const showPhoto = createAction("SHOW_PHOTO");
