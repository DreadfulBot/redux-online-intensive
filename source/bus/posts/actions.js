import { types } from './types';
import { api } from '../../REST';

export const postsActions = {
    fillPosts: posts => {
        return {
            type: types.FILL_POSTS,
            payload: posts
        };
    },
    createPost: post => {
        return {
            type: types.CREATE_POST,
            payload: post
        };
    },

    fetchPostsAsync: () => async dispatch => {
        try {
            dispatch({
                type: types.FETCH_POSTS_ASYNC
            });

            const response = await api.posts.fetch();
            const result = await response.json();

            dispatch(postsActions.fillPosts(result.data));
        } catch (error) {
            dispatch(postsActions.fillPosts([]));
        }
    },

    createPostAsync: message => ({
        type: types.CREATE_POST_ASYNC,
        payload: message
    })
};
