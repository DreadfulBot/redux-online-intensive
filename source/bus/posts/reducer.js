import { fromJS, List } from 'immutable';
import { now } from 'moment';
import { v4 } from 'uuid';
import { mockedProfile } from '../../instruments/mockedData';

import { CREATE_POST_ASYNC, FILL_POSTS } from './types';

const initialState = List();

export const postsReducer = (state = initialState, action) => {
  switch (action.type) {
    case FILL_POSTS:
      return fromJS(action.payload);
    case CREATE_POST_ASYNC:
      let newMessage = {
        id: v4(),
        author: mockedProfile.merge({
          id: v4(),
        }),
        comment: action.payload,
        likes: [],
        profile: mockedProfile.merge({
          id: v4(),
        }),
        created: now(),
      };
      return state.unshift(fromJS(newMessage));
    default:
      return state;
  }
};
