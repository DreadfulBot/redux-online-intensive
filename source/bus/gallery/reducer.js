import { handleActions } from "redux-actions";
import { showNextPhoto, showPhoto, showPrevPhoto } from "./actions";

import photo1 from "../../theme/assets/photos/1.jpeg";
import photo2 from "../../theme/assets/photos/2.jpeg";
import photo3 from "../../theme/assets/photos/3.jpeg";
import photo4 from "../../theme/assets/photos/4.jpeg";

const galleryInitialState = {
  photos: [photo1, photo2, photo3, photo4],
  selectedPhotoIndex: 0,
};

export const gallery = handleActions(
  {
    [showNextPhoto.toString()]: (state, { payload }) => {
      return {
        ...state,
        selectedPhotoIndex:
          (state.selectedPhotoIndex + 1) % state.photos.length,
      };
    },
    [showPrevPhoto.toString()]: (state, { payload }) => {
      return {
        ...state,
        selectedPhotoIndex:
          (state.photos.length + (state.selectedPhotoIndex - 1)) %
          state.photos.length,
      };
    },
    [showPhoto.toString()]: (state, { payload }) => {
      return {
        ...state,
        selectedPhotoIndex: payload % state.photos.length,
      };
    },
  },
  galleryInitialState
);

export default gallery;
