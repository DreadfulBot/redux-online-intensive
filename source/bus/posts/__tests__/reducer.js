import { postsReducer } from "../reducer";
import { postsActions } from "../actions";
import { fromJS } from "immutable";

const initialState = fromJS([__.post]);
const initialStateWithLike = fromJS([{ ...__.post, likes: [__.userProfile]}]);

describe("posts reducer", () => {
    test("should return initial state by default", () => {
        expect(postsReducer(void 0, {})).toMatchSnapshot();
    });

    test("should handle fill posts", () => {
        expect(
            postsReducer(void 0, postsActions.fillPosts([__.post, __.post]))
        ).toMatchSnapshot();
    });

    test("should handle create post", () => {
        expect(
            postsReducer(void 0, postsActions.createPost(__.post))
        ).toMatchSnapshot();
    });

    test("should handle like post", () => {
        expect(
            postsReducer(
                initialState,
                postsActions.likePost({
                    postId: __.postId,
                    liker:  __.userProfile,
                })
            )
        ).toMatchSnapshot();
    });

    test("should handle unlike post", () => {
        expect(
            postsReducer(
                initialStateWithLike,
                postsActions.unlikePost({
                    postId: __.postId,
                    liker:  fromJS(__.userProfile),
                })
            )
        ).toMatchSnapshot();
    });

    test("should handle remove post", () => {
        expect(
            postsReducer(initialState, postsActions.removePost(__.post.id))
        ).toMatchSnapshot();
    });

    test("should handle clear posts", () => {
        expect(
            postsReducer(initialState, postsActions.clearPosts())
        ).toMatchSnapshot();
    });
});
