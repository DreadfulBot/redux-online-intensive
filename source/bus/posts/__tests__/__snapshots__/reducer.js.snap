// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`posts reducer should handle clear posts 1`] = `Immutable.List []`;

exports[`posts reducer should handle create post 1`] = `
Immutable.List [
  Immutable.Map {
    "id": "59edd93052db76aefs2b36ee",
    "comment": "test post",
    "created": 1508760000,
    "likes": Immutable.List [],
    "author": Immutable.Map {
      "id": "TEST_ID",
      "avatar": "TEST_AVATAR",
      "firstName": "Walter",
      "lastName": "White",
      "token": "TEST_TOKEN",
    },
  },
]
`;

exports[`posts reducer should handle fill posts 1`] = `
Immutable.List [
  Immutable.Map {
    "id": "59edd93052db76aefs2b36ee",
    "comment": "test post",
    "created": 1508760000,
    "likes": Immutable.List [],
    "author": Immutable.Map {
      "id": "TEST_ID",
      "avatar": "TEST_AVATAR",
      "firstName": "Walter",
      "lastName": "White",
      "token": "TEST_TOKEN",
    },
  },
  Immutable.Map {
    "id": "59edd93052db76aefs2b36ee",
    "comment": "test post",
    "created": 1508760000,
    "likes": Immutable.List [],
    "author": Immutable.Map {
      "id": "TEST_ID",
      "avatar": "TEST_AVATAR",
      "firstName": "Walter",
      "lastName": "White",
      "token": "TEST_TOKEN",
    },
  },
]
`;

exports[`posts reducer should handle like post 1`] = `
Immutable.List [
  Immutable.Map {
    "id": "59edd93052db76aefs2b36ee",
    "comment": "test post",
    "created": 1508760000,
    "likes": Immutable.List [
      Object {
        "avatar": "TEST_AVATAR",
        "firstName": "Walter",
        "id": "TEST_ID",
        "lastName": "White",
        "token": "TEST_TOKEN",
      },
    ],
    "author": Immutable.Map {
      "id": "TEST_ID",
      "avatar": "TEST_AVATAR",
      "firstName": "Walter",
      "lastName": "White",
      "token": "TEST_TOKEN",
    },
  },
]
`;

exports[`posts reducer should handle remove post 1`] = `Immutable.List []`;

exports[`posts reducer should handle unlike post 1`] = `
Immutable.List [
  Immutable.Map {
    "id": "59edd93052db76aefs2b36ee",
    "comment": "test post",
    "created": 1508760000,
    "likes": Immutable.List [],
    "author": Immutable.Map {
      "id": "TEST_ID",
      "avatar": "TEST_AVATAR",
      "firstName": "Walter",
      "lastName": "White",
      "token": "TEST_TOKEN",
    },
  },
]
`;

exports[`posts reducer should return initial state by default 1`] = `Immutable.List []`;
